### DISCUSSION TITLE
Using Mica on Windows

### DISCUSSION THREAD

**Discussion by StefanKoell** [AvaloniaUI/Avalonia#16091] [Q&A]
Hi,

I asked on Telegram but I guess it passed by too quickly.

I'm trying to figure out how I can create an app on Windows which is using Mica background if possible and as a fall back it should simply use solid colors for background elements. What I can't figure out is how to handle theme changes. There's a property on the Window called `TransparencyBackgroundFallback` (IBrush) but even if I put a DynamicResource in it, it will stick to the initial value - so switching themes do not really work when Mica is not available.

I found a hack which makes it work by setting `TransparencyLevelHint="Mica,Transparent"` and when the `ActualTheme` property changed I do the following on the Window:
```csharp
        TransparencyLevelHint = new[] { WindowTransparencyLevel.Transparent, };
        TransparencyBackgroundFallback = new SolidColorBrush(Colors.Transparent);
        TransparencyBackgroundFallback = ActualThemeVariant == ThemeVariant.Light 
            ? new SolidColorBrush(bgFallbackLight.Color) 
            : new SolidColorBrush(bgFallbackDark.Color);
        TransparencyLevelHint = new[] { WindowTransparencyLevel.Mica, };
```

Since I can't find any docs or samples handling these cases, I'm wondering if I'm doing something wrong of if it is a bug. Has anyone created an app like this before?

Thanks,
Stefan

**Comment by maxkatz6** [comment]
> There's a property on the Window called TransparencyBackgroundFallback (IBrush) but even if I put a DynamicResource in it, it will stick to the initial value

It looks like a bug with TransparencyBackgroundFallback property. Can you open an issue for that?

**Reply by StefanKoell** [reply]
I actually just found out that I am able to make it work as intended by using a style instead. What I do is, I do not set the `Background` of the Window and also do not set the `TransparencyBackgroundFallback`.

Then I just set the `TransparencyLevelHint` either to Mica or to an empty array (through binding, since I also allow the user to disable Mica).

Lastly I created the following style:
```xml
    <Style Selector=":is(Window)[ActualTransparencyLevel=Mica][IsActive=True]">
        <Setter Property="Background" Value="Transparent" />
    </Style>
```

This seems to work. But I'm also not sure if this is the correct way to do it - wondering if it's just a side effect which may stop working at a later point in time.

Note that I explicitly added the IsActive condition to the style because it also seems that Mica is still rendered for inactive windows in Avalonia (maybe another bug).

I also noticed that when Mica rendering is enabled/disabled, it doesn't use an ease-in transition (like on WinUI apps). I tried a brush transition for the Background like this:
```xml
    <Style Selector=":is(Window)[ActualTransparencyLevel=Mica][IsActive=True]">
        <Setter Property="Background" Value="Transparent" />
        <Setter Property="Transitions">
            <Transitions>
                <BrushTransition Property="Background" Duration="0:0:0.5" />
            </Transitions>
        </Setter>
    </Style>
```
... but this doesn't seem to work as it is probably some special treated background in the Win32 implementation.

I will file an issue for the `TransparencyBackgroundFallback` stuff.

Thanks!

**Reply by StefanKoell** [reply]
Bug file here: #16097
